### PR TITLE
Retire legacy printer

### PR DESCRIPTION
**Pull request by tthsqe12**
Please check: acb is expressified, and I changed the padic/qadic printing.

### CODE REVIEW

File: src/arb/acb.jl [hunk 1]
```diff
@@ -180,10 +180,28 @@ end
...
 
-function show(io::IO, x::acb)
-  show(io, real(x))
-  print(io, " + i*")
-  show(io, imag(x))
+function AbstractAlgebra.expressify(z::acb; context = nothing)
+   x = real(z)
+   y = imag(z)
+   if iszero(y) # is exact zero!
+      return AbstractAlgebra.expressify(x)
```
Comment (thofma) [inline src/arb/acb.jl:187]:
```suggestion
      return AbstractAlgebra.expressify(x, context = context)
```

File: src/arb/acb.jl [hunk 2]
```diff
@@ -180,10 +180,28 @@ end
...
+   else
+      y = Expr(:call, :*, AbstractAlgebra.expressify(y), :im)
```
Comment (thofma) [inline src/arb/acb.jl:189]:
```suggestion
      y = Expr(:call, :*, AbstractAlgebra.expressify(y, context = context), :im)
```

File: src/arb/acb.jl [hunk 3]
```diff
@@ -180,10 +180,28 @@ end
...
+      if iszero(x)
+         return y
+      else
+         x = AbstractAlgebra.expressify(x)
```
Comment (thofma) [inline src/arb/acb.jl:193]:
```suggestion
         x = AbstractAlgebra.expressify(x, context = context)
```

File: src/flint/padic.jl [hunk 4]
```diff
@@ -264,33 +264,41 @@ function set_printing_mode(::Type{FlintPadicField}, printing::Symbol)
...
-   p = prime(parent(x))
-
+function AbstractAlgebra.expressify(x::padic; context = nothing)
+   p = BigInt(prime(parent(x)))
    pmode = PADIC_PRINTING_MODE[]
+   sum = Expr(:call, :+)
+   if iszero(x)
+      push!(sum.args, 0)
+   elseif pmode == 0  # terse
+      push!(sum.args, AbstractAlgebra.expressify(lift(FlintQQ, x)))
```
Comment (thofma) [inline src/flint/padic.jl:274]:
```suggestion
      push!(sum.args, AbstractAlgebra.expressify(lift(FlintQQ, x), context = context))
```

File: src/flint/qadic.jl [hunk 5]
```diff
@@ -241,59 +241,39 @@ characteristic(R::FlintQadicField) = 0
...
    c = R()
-   ccall((:padic_poly_get_coeff_padic, libflint), Nothing, (Ref{padic}, Ref{qadic}, Int, Ref{FlintQadicField}), c, x, len, parent(x))
-   while iszero(c)
-     len = len - 1
-     ccall((:padic_poly_get_coeff_padic, libflint), Nothing, (Ref{padic}, Ref{qadic}, Int, Ref{FlintQadicField}), c, x, len, parent(x))
+   for i in degree(parent(b)):-1:0
+      ccall((:padic_poly_get_coeff_padic, libflint), Nothing,
+            (Ref{padic}, Ref{qadic}, Int, Ref{FlintQadicField}),
+            c, b, i, parent(b))
+      ec = expressify(c)
```
Comment (thofma) [inline src/flint/qadic.jl:255]:
```suggestion
      ec = expressify(c, context = context)
```

**Comment by tthsqe12** [comment]
I see all of your suggestions but have a different question: when inside a definition of AA.expressify, why can I just type expressify without the preceding AA.?

**Comment by thofma** [comment]
I guess we did not export expressify from AA. Thus expressify in Nemo refers to Nemo.expressify.

**Comment by fingolfin** [comment]
Perhaps add `import AbstractAlgebra: expressify` and then change `AbstractAlgebra.expressify`  to `expressify`? Seem nicer to me (but of course that's subjective)

**Comment by rfourquet** [comment]
> Perhaps add import AbstractAlgebra: expressify

it's already done, so that the AA prefix in `AbstractAlgebra.expressify` is not necessary, whether when defining the function or just calling it.

**Comment by tthsqe12** [comment]
The docs fail because of AA https://github.com/Nemocas/AbstractAlgebra.jl/blob/master/src/Factor.jl#L101

**Comment by thofma** [comment]
We need first a AA version with this fixed.

**Comment by tthsqe12** [comment]
@thoma I may have gotten the order slightly wrong here.

**Comment by thofma** [comment]
This should be good to go. I checked Hecke and Singular with the recent changes here and in AA, and everything seems to be fine.

**Comment by wbhart** [comment]
Some of the tests have the following error:

```
no method matching needs_parentheses(::fmpz)
```

This seems like it could be relevant.

**Comment by tthsqe12** [comment]
@thofma There is still a slight problem here. The test code does not test the printing of the Fac{T} type, so you wont see it in the tests, but the docs tests do print Fac{fmpz}. The version of AA that nemo currently uses has a printer for Fac{T} that calls needs_parentheses. We could:
1. retore the definition for needs_parentheses(::fmpz), or
2. release new version of AA first (so that printing of Fac{T} does not call needs_parentheses)

**Comment by thofma** [comment]
I will add the `needs_parentheses` back in.

**Comment by wbhart** [comment]
It's kinda weird that only a small number of CI instances failed.

**Comment by tthsqe12** [comment]
only 1.5 tests the docs as part of the tests

**Comment by tthsqe12** [comment]
> I will add the `needs_parentheses` back in.

Technically you would need to define for all of the nemo types since they could all be the T in Fac{T}. ::(:

**Comment by wbhart** [comment]
Ah yeah, I see. Thanks for the explanation @thofma

**Comment by thofma** [comment]
Added the missing `needs_parentheses` and also added some test code for printing of factorizations. Should be good to go if the lights turn green.

**Pull request merged**